File: DAYS/Day54-Group Anagrams/groupAnagram.js
// String - Group Anagrams leetcode 49
// Given an array of strings strs, group the anagrams together. You can return the answer in any order.
// Example 1: Input: strs = ["eat","tea","tan","ate","nat","bat"] Output: [["bat"],["nat","tan"],["ate","eat","tea"]]
// Example 2: Input: strs = [""] Output: [[""]]
// Example 3: Input: strs = ["a"] Output: [["a"]]

// Different Approaches:
// 1. Sorting approach - Time: O(NKlogK) Space: O(NK) 
// 2. Counting approach - Time: O(NK) Space: O(NK)
// Leetcode https://leetcode.com/problems/group-anagrams/solutions/3425535/javascript-sorting-and-counting-approach/
// Solution 2: Counting approach

function groupAnagrams(strs) {
    const map=new Map();
    for(const str of strs){
        const count =new Array(26).fill(0); // Count array 26 letters
        for(const char of str){
            // use charCodeAt to get ASCII code and subtract 'a' to get index 0-25
            count[char.charCodeAt(0) - 'a'.charCodeAt(0)]++ // Increment count for each character
        }
        const key=count.join('#'); // Create a unique key for the count array
        if(!map.has(key)){    
            map.set(key,[]); // If key not present, set empty array
        }
        map.get(key).push(str); //get key letter and push str
    }
    return Array.from(map.values());
}

// Test cases
console.log(groupAnagrams(["eat","tea","tan","ate","nat","bat"])); //Output [["bat"],["nat","tan"],["ate","eat","tea"]]
console.log(groupAnagrams([""])); //Output [[""]]
console.log(groupAnagrams(["a"])); //Output [["a"]]

// Best Approach: Counting approach
// Complexity Analysis: Time Complexity: O(NK) , Space Complexity: O(NK)

File: DAYS/Day54-Group Anagrams/groupAnagram.py
# String - Group Anagrams leetcode 49
# Given an array of strings strs, group the anagrams together. You can return the answer in any order.
# Example 1: Input: strs = ["eat","tea","tan","ate","nat","bat"] Output: [["bat"],["nat","tan"],["ate","eat","tea"]]
# Example 2: Input: strs = [""] Output: [[""]]
# String - Group Anagrams leetcode 49
# Given an array of strings strs, group the anagrams together. You can return the answer in any order.
# Example 1: Input: strs = ["eat","tea","tan","ate","nat","bat"] Output: [["bat"],["nat","tan"],["ate","eat","tea"]]
# Example 2: Input: strs = [""] Output: [[""]]

# Different approaches:
# 1. Sorting Approach: Time O(NKlogK) Space O(NK)
# 2. Counting Approach: Time O(NK) Space O(NK)

# Solution 2: Counting Approach
# Leetcode https://leetcode.com/problems/group-anagrams/solutions/1678829/python-counting-approach-with-explanation/?orderBy=most_relevant
from collections import defaultdict

def groupAnagrams(strs):
    anagram_map = defaultdict(list)
    for word in strs:
        count = [0] * 26  # There are 26 letters in the English alphabet
        for char in word:
            count[ord(char) - ord('a')] += 1  # Increment count alphabet index
        key = tuple(count)
        anagram_map[key].append(word)  # Add word to group
    return list(anagram_map.values())

# Test Cases
if __name__ == "__main__":
    print(groupAnagrams(["eat", "tea", "tan", "ate", "nat", "bat"]))  # Output groups
    print(groupAnagrams([""]))  # Output [[""]]

# Best Approach: Counting Approach
# Complexity Analysis: Time complexity: O(NK) , Space complexity: O(NK)